modules/touch/src/TuioEar.cpp

```diff
@@ -0,0 +1,162 @@
+ * The above copyright notice and this permission notice shall be included in all copies *
+ * or substantial portions of the Software.                                              *
+ *                                                                                       *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,   *
+ * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A         *
+ * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT    *
+ * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF  *
+ * CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE  *
+ * OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                                         *
+ ****************************************************************************************/
+
+#include <modules/touch/include/TuioEar.h>
+
+#include <openspace/engine/openspaceengine.h>
+#include <openspace/engine/settingsengine.h>
+#include <openspace/engine/wrapper/windowwrapper.h>
+#include <openspace/interaction/interactionhandler.h>
+#include <openspace/rendering/renderengine.h>
+#include <openspace/rendering/screenspacerenderable.h>
+
+#include <ghoul/logging/logmanager.h>
+
+using namespace TUIO;
+namespace {
+    const std::string _loggerCat = "TuioEar";
+}
+
+void TuioEar::addTuioObject(TuioObject *tobj) { }
+
+void TuioEar::updateTuioObject(TuioObject *tobj) { }
+
+void TuioEar::removeTuioObject(TuioObject *tobj) { }
+
+void TuioEar::addTuioCursor(TuioCursor *tcur) {
+    _mx.lock();
+    _tap = false;
+    // find same id in _list if it exists in _removeList (new input with same ID as a previously stored)
+    int i = tcur->getSessionID();
+    std::vector<int>::iterator foundID = std::find_if(
+        _removeList.begin(),
+        _removeList.end(),
+        [&i](int id) { return id == i; });
+
+    // if found, remove id from _removeList and update, otherwise add new id to list
+    if (foundID != _removeList.end()) {
+        std::find_if(
+            _list.begin(),
+            _list.end(),
+            [&i](const TuioCursor& cursor) {
+            return cursor.getSessionID() == i;
+        })->update(tcur);
+        _removeList.erase(foundID);
+    }
+    else {
+        _list.push_back(TuioCursor(*tcur));
+    }
+    _mx.unlock();
+}
+
+void TuioEar::updateTuioCursor(TuioCursor *tcur) {
+    _mx.lock();
+    _tap = false;
+    int i = tcur->getSessionID();
+    std::find_if(
+        _list.begin(),
+        _list.end(),
+        [&i](const TuioCursor& cursor) {
+            return cursor.getSessionID() == i;
+    })->update(tcur);
+    _mx.unlock();
+}
+
+// save id to be removed and remove it in clearInput
+void TuioEar::removeTuioCursor(TuioCursor *tcur) {
+    _mx.lock();
+    _removeList.push_back(tcur->getSessionID());
+
+    // Check if the cursor ID could be considered a tap
+    double dist = 0;
```

Comment:
I wish this worked. It seems the `std::list` that is returned from `tcur->getPath()` is a circle-linked list. We never reach `.end()` when I try it, just repeatedly goes through the entire list.